### DISCUSSION TITLE
Deploy PostOwl on Clever Cloud

### DISCUSSION THREAD

**Discussion by juliamrch** [PostOwl/postowl#29] [Show and tell]
## What is Clever Cloud

Clever Cloud is a European cloud provider that handles all the infrastructure so developers can focus on their code. Deployments are immutable, meaning your app will restart when scaling, crashing or updating. PostOwl uses SQLite, so this tutorial explains how  to deploy PostOwl with a replicated and always updated database. At the end, you'll get:

- A deployed PostOwl app
- A really fast integrated database ❤️
- ... continuously replicated! 

For this, we are going to use [Litestream](https://litestream.io) and [Clever Cloud deployment hooks](https://www.clever-cloud.com/doc/develop/build-hooks/) and its S3-like object storage service, [Cellar](https://www.clever-cloud.com/doc/deploy/addon/cellar/).

## Requirements

- Git
- A Clever Cloud account
- [Litestream](https://litestream.io/install/) installed on your machine

## 📁 Backup your local SQLite

You'll need an object storage bucket (also called S3 storage) where you'll store your database snapshots. From the Clever Cloud Console, create an Cellar add-on: 

1. Click on **Create > an add-on**
2. Choose **Cellar**
3. Create a bucket and name it as you wish

On Cellar dashboard, check **Informations** with the environment variables you'll need in your project: 

- `CELLAR_ADDON_HOST`
- `CELLAR_ADDON_KEY_ID`
- `CELLAR_ADDON_KEY_SECRET`

Copy those values and save it somewhere safe.

At the root of your project, create a `litestream.yml` and fill it with these environment variables values:

```yaml
 dbs:
  - path: data/db.sqlite3
    replicas:
      - type: s3
        bucket: <you_bucket_name>
        path: .
        endpoint: <CELLAR_ADDON_HOST>
        region: fr-par
        access-key-id: <CELLAR_ADDON_KEY_ID>
        secret-access-key: <CELLAR_ADDON_KEY_SECRET>
``` 

Now you can run the command to replicate your SQLite database : `./litestream replicate -config litestream.yml`. 

💡I usually advise that you create a specific folder in your bucket to host the database snapshots, you can use [s3cmd CLI](https://s3tools.org/usage) or a S3 client like [Cyberduck](https://cyberduck.io) to do so. If you create a folder in your bucket, make sure to change the `path` value in your `litestream.yml`. 

Finally, add `litestream.yml` to your gitignore, and ⚠️ never commit it anywhere!

## 🛠️ Configure your PostOwl app

At the root of your repository, create a a `litestream.sh` file and fill it with the following: 

```shell
# Download litestream

#!/bin/bash

# Download Litestream
wget https://github.com/benbjohnson/litestream/releases/download/v$LITESTREAM_VERSION/litestream-v$LITESTREAM_VERSION-linux-amd64-static.tar.gz
tar xvzf litestream-v$LITESTREAM_VERSION-linux-amd64-static.tar.gz
chmod +rwx ./litestream


# Check that all necessary environment variables are defined
if [ -z "$DB_PATH" ] || \
   [ -z "$LITESTREAM_BUCKET" ] || \
   [ -z "$BUCKET_DIRECTORY" ] || \
   [ -z "$CELLAR_ADDON_HOST" ] || \
   [ -z "$CELLAR_REGION" ] || \
   [ -z "$CELLAR_ADDON_KEY_ID" ] || \
   [ -z "$CELLAR_ADDON_KEY_SECRET" ]; then
  echo "Some environment variables are missing. Please set them before running this script."
  exit 1
fi

# Create the YAML file
cat <<EOL > $APP_HOME/litestream.yml
dbs:
  - path: $DB_PATH
    replicas:
      - type: s3
        bucket: $LITESTREAM_BUCKET
        path: $LITESTREAM_BACKUPS
        endpoint: $CELLAR_ADDON_HOST
        region: $CELLAR_REGION
        access-key-id: $CELLAR_ADDON_KEY_ID
        secret-access-key: $CELLAR_ADDON_KEY_SECRET
EOL

echo "litestream.yml file created successfully."

# Give appropriate permissions
chmod +rwx litestream.yml

# Restore last SQLite backup
rm -rf $DATA_DIR
mkdir -p $DATA_DIR
./litestream restore -config litestream.yml data/db.sqlite3
```
This file will:

- Download Litestream on Clever Cloud server
- Create a configuration file for Litestream
- Remove your database and recreate it with the last backups from your Cellar bucket

Give it permissions with `chmod +rwx litestream.sh` and commit it.

Time to deploy! 

## 🚀 Deployment

Follow these steps to deploy your PostOwl app.

### 1. Declare your app

1. From Clever Cloud console, click on **Create>an application**
5. Choose your deployment method (Git or Github)
6. Choose a **NodeJS**  runtime and select your instance size (smallest one works for a small project)
7. Inject the following environment variables (use **Expert**)

```
BUCKET_DIRECTORY= "."
CC_PRE_RUN_HOOK="./litestream.sh"
CC_WORKER_COMMAND="./litestream replicate -config litestream.yml"
CC_WORKER_RESTART="always"
CC_WORKER_RESTART_DELAY="60"
CELLAR_REGION="fr-par"
DATA_DIR="data"
DB_PATH="data/db.sqlite3"
LITESTREAM_BACKUPS="."
LITESTREAM_BUCKET="<name-of-your-bucket>"
LITESTREAM_VERSION="0.3.9"
```
⚠️ If you've created a folder in your Cellar bucket, make sure to change the values of `BUCKET_DIRECTORY` and `LITESTREAM_BACKUPS`.
 
Add variables from your project's `.env` file as well. Variables will be dynamically injected on deployment.

### 2. Connect Cellar add-on

Before pushing your code, we'll connect your Cellar add-on to your app so its environment variables will be injected as well.

1. From your NodeJS app menu, click on **Service dependencies**
2. On **Link add-ons**, select your Cellar add-on

### 3. Push the code

1. From your NodeJS app menu, go to **Information**
2. Copy the deployment command
3. Paste it at the root of your project and launch it

Now your deployment has started, which you can follow from the **Logs** panel.

**Comment by keybits** [comment]
@juliamrch Thank you so much, this is brilliant.

I haven't tried it yet, but will give it a go when I get back from being on the road.

Couple of questions:

1. How much does it cost per month to host PostOwl this way?
2. Would it be OK for me to add these instructions to the [PostOwl docs](https://www.postowl.com/docs/administration-guide/deploy/)?

**Reply by juliamrch** [reply]
Hi @keybits, my pleasure! PostOwl will serve to a lot of people around me who just want to blog and basta.

**1. Regarding the cost:** I've tried the smallest NodeJS instance (pico) and restarted the server and it behaves very well, it costs 4.50 €/month (less than 5 $ right now). I would usually enable auto scalability just to be safe if I had traffic on it, it's billed per second, just the time I'd need more ressources. But right now, if you look at the server metrics after restarting the app, Pico is enough for a demo blog no one is visiting and I got some margin.

<img width="918" alt="overview" src="https://github.com/PostOwl/postowl/assets/101819212/a7933246-7302-4a35-bb2b-28a999524426">

For the build, however, since it usually takes more ressources (and I'm terrified of npm for that) I enable a custom build instance just in case. Again, It's billed per second. I don't know if I would try to build on a Pico instance on production. Maybe if I'm feeling _kamikaze_. And it's not a big deal, because if the deployment fails, the app is still up, the previous instance keeps on running like usual.

<img width="897" alt="build" src="https://github.com/PostOwl/postowl/assets/101819212/53d17776-a1ea-4592-9441-35202d8b785d">

**2. Including this in PostOwl docs:** sure, be my guest !
